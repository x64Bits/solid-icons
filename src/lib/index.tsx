import { JSX, splitProps } from "solid-js";
import { isServer, mergeProps, ssr } from "solid-js/web";

type SVGSVGElementTags = JSX.SVGElementTags["svg"];

export interface IconTree {
  a: SVGSVGElementTags;
  c: string;
}

export interface IconProps extends SVGSVGElementTags {
  size?: string | number;
  color?: string;
  title?: string;
  style?: JSX.CSSProperties;
}

export interface IconBaseProps extends IconProps {
  src: IconTree;
}

export declare type IconTypes = (props: IconProps) => JSX.Element;

export const CustomIcon = (props: IconBaseProps) =>
  IconTemplate(props.src, props);

export function IconTemplate(iconSrc: IconTree, props: IconProps): JSX.Element {
  const mergedProps = mergeProps(iconSrc.a, props) as IconBaseProps;
  const [_, svgProps] = splitProps(mergedProps, ["src"]);

  return (
    <svg
      stroke={iconSrc.a.stroke}
      color={props.color || "currentColor"}
      stroke-width="0"
      style={{
        ...props.style,
        overflow: "visible",
      }}
      {...svgProps}
      height={props.size || "1em"}
      width={props.size || "1em"}
      innerHTML={iconSrc.c}
      xmlns="http://www.w3.org/2000/svg"
    >
      {isServer && ssr(iconSrc.c)}
      {props.title && <title>{props.title}</title>}
    </svg>
  );
}
