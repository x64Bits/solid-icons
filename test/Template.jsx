import { createSignal } from "solid-js";
import { afterEach, describe, expect, test } from "vitest";
import { render, fireEvent, screen, cleanup } from "solid-testing-library";

import { IconTemplate } from "../src/lib/index";

const exampleIcon = {
  a: { fill: "currentColor", viewBox: "0 0 16 16" },
  c: '<path d="M2 0a2 2 0 00-2 2v12a2 2 0 002 2h12a2 2 0 002-2V2a2 2 0 00-2-2H2zm7.283 4.002V12H7.971V5.338h-.065L6.072 6.656V5.385l1.899-1.383h1.312z"/>',
};

const Icon = (props) => IconTemplate(exampleIcon, props);
const initialColor = "#2c4f7c";

describe("<IconTemplate />", () => {
  afterEach(cleanup);

  test("render child", () => {
    const { container, unmount } = render(() => <Icon />);
    const children = container.getElementsByTagName("path");

    expect(children).toHaveLength(1);
    unmount();
  });

  test("reactive props", async () => {
    const expectedColor = "#0CDC73";

    const [color, setColor] = createSignal(initialColor);

    render(() => (
      <Icon
        color={color()}
        onClick={() => setColor(expectedColor)}
        role="svg"
      />
    ));

    const svg = await screen.findByRole("svg");

    fireEvent.click(svg);
    await Promise.resolve();

    const elementColor = svg.getAttribute("color").toLowerCase();
    expect(elementColor).toMatchSnapshot(`#${expectedColor}`);
  });

  test("viewBox defined", async () => {
    render(() => <Icon role="svg" />);
    const svg = await screen.findByRole("svg");

    const viewBox = svg.getAttribute("viewBox");
    expect(viewBox).toBe(exampleIcon.a.viewBox);
  });

  test("default values", async () => {
    render(() => <Icon role="svg" />);
    const svg = await screen.findByRole("svg");

    const defaultHeight = svg.getAttribute("height");
    const defaultColor = svg.getAttribute("color");

    expect(defaultHeight).toBe("1em");
    expect(defaultColor).toBe("currentColor");
  });

  test("custom values", async () => {
    const customSize = "3em";

    render(() => <Icon role="svg" size={customSize} color={initialColor} />);
    const svg = await screen.findByRole("svg");

    const customHeight = svg.getAttribute("height");
    const customWidth = svg.getAttribute("width");
    const customColor = svg.getAttribute("color");

    expect(customHeight).toBe(customSize);
    expect(customWidth).toBe(customSize);
    expect(customColor).toBe(initialColor);
  });
});
