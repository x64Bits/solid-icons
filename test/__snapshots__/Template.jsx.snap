// Vitest Snapshot v1

exports[`<IconTemplate /> > reactive props > ##0CDC73 1`] = `"#0cdc73"`;

exports[`<IconTemplate /> > reactive props > #0CDC73 1`] = `"0cdc73"`;
